File: hitl-vsc-extension/server/app/schemas.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class Message(BaseModel):
    role: str
    content: str


class AgentRequestEnvelope(BaseModel):
    student_id: str
    session_id: str = Field(json_schema_extra={"format": "uuid"})
    project_id: str
    arm_id: Literal[1, 2, 3]
    messages: list[Message]
    timestamp: datetime
    workflow_landmark: str | None
    planning_document: str | None
    client_version: str


class InstructionalResponseEnvelope(BaseModel):
    response_content: str
    drift_injected: bool
    intervention_id: str | None
    original_response_hash: str | None
    server_timestamp: datetime


class AuthRegisterRequest(BaseModel):
    student_token: str


class LabConfig(BaseModel):
    landmarks: list[str]


class AuthRegisterResponse(BaseModel):
    arm_id: Literal[1, 2, 3]
    project_id: str
    session_id: str
    lab_config: LabConfig

File: clients/vscode/extension/src/schemas.ts
import { z } from "zod";

const MessageSchema = z.object({
  role: z.string(),
  content: z.string(),
});

export const AgentRequestEnvelopeSchema = z.object({
  student_id: z.string(),
  session_id: z.string().uuid(),
  project_id: z.string(),
  arm_id: z.union([z.literal(1), z.literal(2), z.literal(3)]),
  messages: z.array(MessageSchema),
  timestamp: z.string().datetime(),
  workflow_landmark: z.string().nullable(),
  planning_document: z.string().nullable(),
  client_version: z.string(),
});

export type AgentRequestEnvelope = z.infer<typeof AgentRequestEnvelopeSchema>;

export const InstructionalResponseEnvelopeSchema = z.object({
  response_content: z.string(),
  drift_injected: z.boolean(),
  intervention_id: z.string().nullable(),
  original_response_hash: z.string().nullable(),
  server_timestamp: z.string().datetime(),
});

export type InstructionalResponseEnvelope = z.infer<
  typeof InstructionalResponseEnvelopeSchema
>;

export const LabConfigSchema = z.object({
  landmarks: z.array(z.string()),
});

export type LabConfig = z.infer<typeof LabConfigSchema>;

export const AuthRegisterResponseSchema = z.object({
  arm_id: z.union([z.literal(1), z.literal(2), z.literal(3)]),
  project_id: z.string(),
  session_id: z.string().uuid(),
  lab_config: LabConfigSchema,
});

export type AuthRegisterResponse = z.infer<
  typeof AuthRegisterResponseSchema
>;
